### PR TITLE
Update resource-name generation to match the revised specification

### PR DESCRIPTION
**Pull request by chrisdunelm**

### CODE REVIEW

File: Google.Api.Generator.Tests/ProtoTests/ChildResource/Testing.ChildResource/ChildResourceClient.g.cs [hunk 1]
```diff
@@ -16,226 +16,2486 @@
...
 
+        /// <summary>
+        /// </summary>
+        /// <param name="ref">
+        /// </param>
+        /// <param name="repeatedRef">
+        /// </param>
+        /// <param name="callSettings">If not null, applies overrides to this RPC call.</param>
+        /// <returns>A Task containing the RPC response.</returns>
+        public virtual stt::Task<Response> TripleParentMethodAsync(ProjectName @ref, scg::IEnumerable<gax::IResourceName> repeatedRef, gaxgrpc::CallSettings callSettings = null) =>
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/ChildResource/Testing.ChildResource/ChildResourceClient.g.cs]: I can't remember why we have IResourceName overloads here - a TripleName parent is always one of ProjectName, DeptName, OrgName... do we always add an IResourceName overload whenever child_type is specified? Or only when there are multiple parents? (I'd expect this for TripleWildcard, but not TripleParent. I suspect there's a good reason for it - I don't have the doc open right now.)

File: Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNamesFakes.cs [hunk 2]
```diff
@@ -12,44 +12,78 @@
...
     }
 
-    public partial class OriginallySingleResourceRef : ProtoMsgFake<OriginallySingleResourceRef>
-    {
-        public string Resource1 { get; set; }
-        public string Resource2 { get; set; }
-    }
+    public partial class Response : ProtoMsgFake<Response> { }
+
+    //public partial class SingleResource : ProtoMsgFake<SingleResource>
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNamesFakes.cs]: Can these be removed now? Not sure why they're here.

File: Google.Api.Generator.Tests/ProtoTests/ChildResource/Testing.ChildResource/ChildResourceClient.g.cs (hunk 1 above)
Comment (chrisdunelm) [reply Google.Api.Generator.Tests/ProtoTests/ChildResource/Testing.ChildResource/ChildResourceClient.g.cs]: Yes, you're right, these shouldn't be here. The `IResourceName` properties are required in the message partial class, but not as a method overload here. I'll fix this.

File: Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNamesFakes.cs (hunk 2 above)
Comment (chrisdunelm) [reply Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNamesFakes.cs]: Oops, yes, an oversight. Will fix.

**Comment by chrisdunelm** [comment]
String pagination is still supported, but for now it doesn't attempt to paginate over a resource-name, it just paginates over the string.
I'd quite like to leave it this way in this PR, and file an issue to work out how to handle this long-term.

For the three affected APIs:
* Pubsub - [`ListTopicSnapshotsResponse`](https://github.com/googleapis/googleapis/blob/b491d07cadaae7cde5608321f913e5ca1459b32d/google/pubsub/v1/pubsub.proto#L298) and [`ListTopicSubscriptionsResponse`](https://github.com/googleapis/googleapis/blob/b491d07cadaae7cde5608321f913e5ca1459b32d/google/pubsub/v1/pubsub.proto#L272). Both will be resource annotated.
* Logging - [`ListLogsResponse`](https://github.com/googleapis/googleapis/blob/b491d07cadaae7cde5608321f913e5ca1459b32d/google/logging/v2/logging.proto#L351). This will be resource annotated.
* Firestore - [`ListCollectionIdsResponse`](https://github.com/googleapis/googleapis/blob/b491d07cadaae7cde5608321f913e5ca1459b32d/google/firestore/v1/firestore.proto#L755). This is *not* a resource.

File: Google.Api.Generator.Tests/ProtoTests/ChildResource/Testing.ChildResource/ChildResourceClient.g.cs (hunk 1 above)
Comment (chrisdunelm) [reply Google.Api.Generator.Tests/ProtoTests/ChildResource/Testing.ChildResource/ChildResourceClient.g.cs]: Just thinking that the partial class `IResourceName` properties should probably be `get`-only if there isn't a wildcard specified. When setting the user should always have a strong-typed resource-name; it's only when getting that the type can't be known.

**Comment by chrisdunelm** [comment]
Comments now addressed. PTAL.
Apologies about the unpleasantness of reviewing this PR - as usual, please do look at the generated code, especially for the ResourceName and ChildResource tests - mostly that all the correct signatures are present, and the correct partial-proto-message-class properties are present.

File: Google.Api.Generator.Tests/ProtoTests/ChildResource/Testing.ChildResource/ChildResourceClient.g.cs (hunk 1 above)
Comment (jskeet) [reply Google.Api.Generator.Tests/ProtoTests/ChildResource/Testing.ChildResource/ChildResourceClient.g.cs]: Just to clarify this conversation, Chris and I then chatted offline - one common scenario where you might *not* know the type is copying from one object to another, which is why Chris re-added the setter.

File: Google.Api.Generator.Tests/ProtoTests/Paginated/Testing.Paginated/PaginatedClient.g.cs [hunk 3]
```diff
@@ -209,8 +209,8 @@ public abstract class PaginatedClient
         /// </summary>
         /// <param name="request">The request object containing all of the parameters for the API call.</param>
         /// <param name="callSettings">If not null, applies overrides to this RPC call.</param>
-        /// <returns>A pageable sequence of <see cref="ResourceName"/> resources.</returns>
-        public virtual gax::PagedEnumerable<ResourceResponse, ResourceName> ResourcedMethod(ResourceRequest request, gaxgrpc::CallSettings callSettings = null) =>
+        /// <returns>A pageable sequence of <see cref="string"/> resources.</returns>
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/Paginated/Testing.Paginated/PaginatedClient.g.cs]: It seems odd to see a string referred to as a resource, but it's probably not worth special-casing it.

File: Google.Api.Generator/ProtoUtils/ResourceDetails.cs [hunk 4]
```diff
@@ -174,54 +204,92 @@ string GetShortName(ResourceDescriptor resDesc)
...
-                // It is invalid to ask for a parent that is not already defined.
-                // This may change in the future, to allow auto-generating parent resource-names, but this is not currently allowed.
-                throw new InvalidOperationException(
-                    $"Cannot refer to child-type '{resourceRef.ChildType}' in field {msgDesc.Name}.{fieldDesc.Name} because the child patterns are not already defined in a resource.");
             }
-            throw new InvalidOperationException("type or child_type must be set.");
 
             string ParentPattern(string pattern)
             {
+                if (pattern == "*")
```
Comment (jskeet) [inline Google.Api.Generator/ProtoUtils/ResourceDetails.cs]: Ooh, is this something I forgot to do in a previous PR? Sorry.

File: Google.Api.Generator/RoslynUtils/RoslynBuilder.cs [hunk 5]
```diff
@@ -167,14 +167,15 @@ public static PropertyDeclarationSyntax AutoProperty(Modifier modifiers, TypeSyn
...
-                SimpleLambdaExpression(parameter.WithType(null), Block(ToStatements(code).ToArray()));
+            var statements = ToStatements(code).ToArray();
+            var expr = SimpleLambdaExpression(parameter.WithType(null), statements.Length == 1 ? MakeExpr(statements[0]) : Block(statements));
             if (async)
             {
                 expr = expr.WithAsyncKeyword(Token(SyntaxKind.AsyncKeyword));
             }
             return expr;
+            CSharpSyntaxNode MakeExpr(CSharpSyntaxNode c) =>
+                c is ReturnStatementSyntax ret ? ret.Expression : c is ExpressionStatementSyntax exprState ? exprState.Expression : c;
```
Comment (jskeet) [inline Google.Api.Generator/RoslynUtils/RoslynBuilder.cs]: I'd break after each colon here, so it's a sequence of "condition ? value : " lines... effectively like a switch expression. (Presumably you'd use a switch expression here if you were on C# 8...)

File: Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNames.proto [hunk 6]
```diff
@@ -9,97 +9,69 @@ import "google/api/resource.proto";
...
   }
 }
 
-message SingleResource {
+message SinglePattern {
   option (google.api.resource) = {
-    type: "resourcenames.example.com/SingleResource"
+    type: "resourcenames.example.com/SinglePattern"
 	pattern: "items/{item_id}"
+    name_field: "real_name"
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNames.proto]: Is there a reason why pattern is more indented than type and name_field?

File: Google.Api.Generator.Tests/ProtoTests/Paginated/Testing.Paginated/PaginatedClient.g.cs (hunk 3 above)
Comment (chrisdunelm) [reply Google.Api.Generator.Tests/ProtoTests/Paginated/Testing.Paginated/PaginatedClient.g.cs]: ack

File: Google.Api.Generator/ProtoUtils/ResourceDetails.cs (hunk 4 above)
Comment (chrisdunelm) [reply Google.Api.Generator/ProtoUtils/ResourceDetails.cs]: I don't think so, it was the simplest way to get multi-parent-including-wildcard-patterns working ;)

File: Google.Api.Generator/RoslynUtils/RoslynBuilder.cs (hunk 5 above)
Comment (chrisdunelm) [reply Google.Api.Generator/RoslynUtils/RoslynBuilder.cs]: done

File: Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNames.proto (hunk 6 above)
Comment (chrisdunelm) [reply Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNames.proto]: The whole tabs/spaces thing. Fixed.

**Comment by chrisdunelm** [comment]
OK, everything ready for merging now, I'll leave that to you...

**Pull request merged**